src/promql/src/extension_plan/series_divide.rs

```diff
@@ -255,22 +255,27 @@ impl Stream for SeriesDivideStream {
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         loop {
-            // It has to be cloned here, otherwise the later ready! will mess things up
-            if let Some(batch) = self.buffer.clone() {
+            if let Some(batch) = self.buffer.as_ref() {
                 let same_length = self.find_first_diff_row(&batch) + 1;
                 if same_length >= batch.num_rows() {
                     let next_batch = match ready!(self.as_mut().fetch_next_batch(cx)) {
-                        Some(Ok(batch)) => batch,
+                        Some(Ok(next_batch)) => next_batch,
                         None => {
-                            self.buffer = None;
                             self.num_series += 1;
-                            return Poll::Ready(Some(Ok(batch)));
+                            return Poll::Ready(self.buffer.take().map(Ok));
                         }
                         error => return Poll::Ready(error),
                     };
-                    let new_batch =
-                        compute::concat_batches(&batch.schema(), &[batch.clone(), next_batch])?;
-                    self.buffer = Some(new_batch);
+                    match self.buffer.take() {
+                        None => {
+                            self.buffer = Some(next_batch);
```

Comment:
This is the same as in the `self.buffer.as_ref matches None` branch. Somehow reasonable.

Reply:
Is it possible to reach this branch? `if-let` should ensure `self.buffer` is always `Some`.

Reply:
> Is it possible to reach this branch?

Currently no. But the `if-let` condition is a loosely guard. Said if we change `buffer` future in `fetch_next_batch` (just as time goes, developers forget why we write the code), such `unwrap` MAY crash.

We can add this branch as defensive programming since it's semantically correct. If it's not reached, no downside (`.unwrap` will check if None also).

Reply:
I push a new commit to tidy the code so that it's not longy.

But I agree that if we get `None` here, the state may be corrupted and we may need panic.

Reply:
> if we get None here, the state may be corrupted and we might need panic.

Yes. I'd prefer to use unwrap.

> But the if-let condition is a loosely guard. Said if we change buffer future in fetch_next_batch (just as time goes, developers forget why we write the code), such unwrap MAY crash.

This case should be covered by a unit test, if possible. We can't guarantee what we do in the `None` branch still works correctly if we change the buffer future.

Reply:
+1 for document and tests... It would be easy to break this in future changes.

Reply:
Well, it looks hard to cover this branch...

Reply:
Push a new commit to use `except`. Please take a look @tisonkun @evenyag

Reply:
https://github.com/GreptimeTeam/greptimedb/pull/3543/commits/9dc00c150bd93cc5c0ceb94bf86287931d2a2723